File: backend/agents/graph.js
class SharedGraph {
  constructor(asset) {
    this.asset = asset;
    this.nodes = new Map();
    this.edges = new Map();
    this.signals = [];
    this.createdAt = Date.now();
    this.addNode({ id: 'root', type: 'asset', label: asset.symbol, category: 'root', data: { name: asset.name, assetType: asset.assetType }, weight: 1.0 });
  }

  addNode({ id, type, label, category, data = {}, weight = 0.5, agentId = null, agentName = null }) {
    this.nodes.set(id, { id, type, label, category, data, weight, agentId, agentName, ts: Date.now() });
  }

  addEdge({ id, source, target, label = '', weight = 0.5, direction = 'neutral', type = 'signal' }) {
    this.edges.set(id, { id, source, target, label, weight, direction, type, ts: Date.now() });
  }

  addSignal({ agentId, agentName, role, roleName, signal, confidence, reasoning, keyFactors, butterflies }) {
    this.signals.push({ agentId, agentName, role, roleName, signal, confidence, reasoning, keyFactors, butterflies, ts: Date.now() });
  }

  toSynthesizerInput() {
    const nodes = Array.from(this.nodes.values()).filter(n => n.type !== 'asset');
    const edges = Array.from(this.edges.values()).filter(e => e.type === 'causal');
    const bull = this.signals.filter(s => s.signal === 'bullish').length;
    const bear = this.signals.filter(s => s.signal === 'bearish').length;
    const neut = this.signals.filter(s => s.signal === 'neutral').length;
    const total = this.signals.length || 1;
    const avgConf = this.signals.length ? Math.round(this.signals.reduce((s, a) => s + (a.confidence || 0), 0) / this.signals.length) : 0;

    const nodeText = nodes.map(n => {
      const d = n.data || {};
      return `[${n.id}] ${n.label} (${n.category}) agent="${n.agentName || '?'}" conf=${d.confidence || 0}% signal=${d.signal || '?'}\n  Reasoning: ${String(d.reasoning || '').slice(0, 300)}\n  KeyFactors: ${(d.keyFactors || []).join(' | ')}\n  Butterflies: ${(d.butterflies || []).join(' | ')}`;
    }).join('\n\n');

    const edgeText = edges.map(e => `  ${e.source} --[${e.direction}/${e.weight.toFixed(2)}]--> ${e.target} "${e.label}"`).join('\n');
    const voteText = this.signals.map(s => `  ${s.roleName} (${s.agentName}): ${s.signal} @ ${s.confidence}%`).join('\n');

    return `ASSET: ${this.asset.symbol} (${this.asset.name || ''})
DATE: ${new Date().toISOString().split('T')[0]}
TOTAL AGENTS: ${this.signals.length} | NODES: ${this.nodes.size} | EDGES: ${this.edges.size}

VOTE TALLY:
  Bullish: ${bull}/${total} (${Math.round(bull/total*100)}%)
  Bearish: ${bear}/${total} (${Math.round(bear/total*100)}%)
  Neutral: ${neut}/${total} (${Math.round(neut/total*100)}%)
  Avg confidence: ${avgConf}%

INDIVIDUAL VOTES:
${voteText}

AGENT NODES (full analysis):
${nodeText}

CAUSAL EDGE CLAIMS (butterfly chains declared by agents):
${edgeText || '  (none)'}`;
  }

  toFlowFormat() {
    const bull = this.signals.filter(s => s.signal === 'bullish').length;
    const bear = this.signals.filter(s => s.signal === 'bearish').length;
    const neut = this.signals.filter(s => s.signal === 'neutral').length;
    return {
      nodes: Array.from(this.nodes.values()),
      edges: Array.from(this.edges.values()),
      signals: this.signals,
      stats: { totalNodes: this.nodes.size, totalEdges: this.edges.size, totalAgents: this.signals.length, bull, bear, neut }
    };
  }
}

module.exports = { SharedGraph };

File: backend/app/services/graph.py
import time
from dataclasses import dataclass, field, asdict
from typing import Optional

@dataclass
class Node:
    id: str; type: str; label: str; weight: float=0.5
    agent_id: Optional[str]=None; data: dict=field(default_factory=dict)
    def to_dict(self): return asdict(self)

@dataclass
class Edge:
    id: str; source: str; target: str
    label: str=''; direction: str='neutral'; weight: float=0.5; type: str='link'
    def to_dict(self): return asdict(self)

@dataclass
class Vote:
    agent_id: str; agent_name: str; role: str
    signal: str; confidence: int; reasoning: str
    factors: list=field(default_factory=list); butterflies: list=field(default_factory=list)
    def to_dict(self): return asdict(self)

class IntelGraph:
    def __init__(self, sym, asset_type):
        self.sym=sym; self.asset_type=asset_type
        self.nodes={}; self.edges={}; self.votes=[]
        self.nodes['root']=Node(id='root',type='asset',label=sym,weight=1.0)

    def add_node(self, n: Node): self.nodes[n.id]=n
    def add_edge(self, e: Edge): self.edges[e.id]=e
    def add_vote(self, v: Vote): self.votes.append(v)

    @property
    def stats(self):
        b=sum(1 for v in self.votes if v.signal=='bullish')
        be=sum(1 for v in self.votes if v.signal=='bearish')
        n=sum(1 for v in self.votes if v.signal=='neutral')
        return {'nodes':len(self.nodes),'edges':len(self.edges),'agents':len(self.votes),
                'bull':b,'bear':be,'neut':n}

    def to_synth_prompt(self):
        st=self.stats; t=st['agents'] or 1
        votes='\n'.join(f"  {v.role}({v.agent_name}): {v.signal} @{v.confidence}%" for v in self.votes)
        agents='\n\n'.join(
            f"[{n.id}] {n.label}({n.type}) signal={n.data.get('signal')} conf={n.data.get('confidence')}%\n"
            f"  Reasoning: {str(n.data.get('reasoning',''))[:300]}\n"
            f"  Factors: {' | '.join(n.data.get('factors',[]))}\n"
            f"  Butterflies: {' | '.join(n.data.get('butterflies',[]))}"
            for n in self.nodes.values() if n.type=='specialist')
        causal='\n'.join(
            f"  {e.source} --[{e.direction}]--> {e.target} '{e.label}'"
            for e in self.edges.values() if e.type=='causal') or '  (none)'
        return (f"ASSET: {self.sym} ({self.asset_type}) DATE: {time.strftime('%Y-%m-%d')}\n"
                f"VOTE TALLY: {st['bull']}B / {st['bear']}Be / {st['neut']}N from {t} agents\n\n"
                f"AGENT SIGNALS:\n{votes}\n\nAGENT NODES:\n{agents}\n\nCAUSAL CHAINS:\n{causal}")

    def to_flow(self):
        return {'nodes':[n.to_dict() for n in self.nodes.values()],
                'edges':[e.to_dict() for e in self.edges.values()],
                'votes':[v.to_dict() for v in self.votes],'stats':self.stats}
